src/compiler/common_sub_expression.cc

```diff
@@ -0,0 +1,57 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file common_sub_expression.cc
+ * \brief take common expressions out and replace with one.
+ *
+ */
+#include <nnvm/graph.h>
+#include <nnvm/pass.h>
+
+namespace nnvm {
+namespace compiler {
+
+nnvm::Graph CommonSubExpression(nnvm::Graph src) {
+  std::vector<NodeEntry> commonNodeList;
+  std::unordered_map<std::string, int> unique_expressn;
+  std::string expressn = "";
+  std::function<std::string(const NodeEntry& node, std::string)>find_cummulative_expression
+    = [&find_cummulative_expression, &commonNodeList, &unique_expressn]
+    (const NodeEntry& node, std::string expressn)->std::string {
+    std::string tempExprssn = "";
+    if (node.node->is_variable()) {
+      tempExprssn += node.node->attrs.name;
+    } else {
+      for (auto& e : node.node->inputs) {
+        tempExprssn = find_cummulative_expression(e, tempExprssn);
+        // Replace already commoned expression with its global index
+        if (unique_expressn.count(tempExprssn)) {
+          e = commonNodeList[unique_expressn.at(tempExprssn)];
+          tempExprssn = std::to_string(unique_expressn.at(tempExprssn));
+        }
+        expressn += tempExprssn;
+        tempExprssn = "";
+      }
+
+      tempExprssn = node.node->op()->name + expressn;
```

Comment:
what if nodes have the same op but different attribute?

Reply:
@ZihengJiang , this is not handled in current code, i will modify code and update the PR. Thanks for the comment!